PsfLauncher/main.cpp

```diff
@@ -13,223 +13,227 @@
 #include <psf_runtime.h>
 #include <shellapi.h>
 #include <combaseapi.h>
+#include <ppltasks.h>
+#include <ShObjIdl.h>
+#include "ErrorInformation.h"
 
-using namespace std::literals;
+//These two macros don't exist in RS1.  Define them here to prevent build
+//failures when building for RS1.
+#ifndef PROCESS_CREATION_DESKTOP_APP_BREAKAWAY_DISABLE_PROCESS_TREE
+#    define PROCESS_CREATION_DESKTOP_APP_BREAKAWAY_DISABLE_PROCESS_TREE PROCESS_CREATION_DESKTOP_APPX_OVERRIDE
+#endif
 
-// Forward declaration
-void LaunchMonitorInBackground(std::filesystem::path packageRoot, const wchar_t * executable, const wchar_t * arguments, bool wait, bool asadmin);
+#ifndef PROC_THREAD_ATTRIBUTE_DESKTOP_APP_POLICY
+#    define PROC_THREAD_ATTRIBUTE_DESKTOP_APP_POLICY PROC_THREAD_ATTRIBUTE_CHILD_PROCESS_POLICY
+#endif
 
-static inline bool check_suffix_if(iwstring_view str, iwstring_view suffix)
-{
-    if ((str.length() >= suffix.length()) && (str.substr(str.length() - suffix.length()) == suffix))
-    {
-        return true;
-    }
-
-    return false;
-}
+using namespace std::literals;
 
-void Log(const char* fmt, ...)
+struct ExecutionInformation
 {
-    std::string str;
-    str.resize(256);
-
-    va_list args;
-    va_start(args, fmt);
-    std::size_t count = std::vsnprintf(str.data(), str.size() + 1, fmt, args);
-    assert(count >= 0);
-    va_end(args);
-
-    if (count > str.size())
-    {
-        str.resize(count);
-
-        va_list args2;
-        va_start(args2, fmt);
-        count = std::vsnprintf(str.data(), str.size() + 1, fmt, args2);
-        assert(count >= 0);
-        va_end(args2);
-    }
+    LPCWSTR ApplicationName;
+    std::wstring CommandLine;
+    LPCWSTR CurrentDirectory;
+};
+
+//Forward declarations
+ErrorInformation StartProcess(ExecutionInformation execInfo, int cmdShow, bool runInAppContainer);
+int launcher_main(PWSTR args, int cmdShow) noexcept;
+ErrorInformation RunScript(const psf::json_object * scriptInformation, std::filesystem::path packageRoot, LPCWSTR dirStr, int cmdShow);
+ErrorInformation GetAndLaunchMonitor(const psf::json_object * monitor, std::filesystem::path packageRoot, int cmdShow, LPCWSTR dirStr);
+ErrorInformation LaunchMonitorInBackground(std::filesystem::path packageRoot, const wchar_t * executable, const wchar_t * arguments, bool wait, bool asadmin, int cmdShow, LPCWSTR dirStr);
+static inline bool check_suffix_if(iwstring_view str, iwstring_view suffix);

```

Comment:
lots of pointers - use references where possible

Reply:
There are a lot of pointers.  I changed the C-style strings from * to [] but I am not going to change the json_objects to references.  At least, not for this PR.  The reason why is config.cpp returns pointers.  I could make them return a reference instead, but what would I return in case the object could not be found?

I could just make an empty json_object, but I don't know how that would affect the rest of the program.

So, right now, I won't change it.  But, that is something I can change in the future.

Reply:
I'm referring to what a function *takes* (not what it returns).  If it never expects a null value, than reference is a stronger contract and avoids the risk, param validation, etc.